src/app/GitUI/Editor/Diff/DiffHighlightService.cs

```diff
@@ -126,115 +127,137 @@ public override void AddTextHighlighting(IDocument document)
+        }
 
-        for (;
-            line < document.TotalNumberOfLines
-            && DoesLineStartWith(document, endLine.Offset, prefixStr, invertMatch);
-            line++)
+        foreach (ISegment segment in GetAllLines(DiffLineType.Header))
         {
-            endLine = document.GetLineSegment(line);
+            document.MarkerStrategy.AddMarker(CreateTextMarker(segment.Offset, segment.Length, AppColor.DiffSection.GetThemeColor()));
         }
+    }
 
-        line = Math.Max(0, line - 2);
-        endLine = document.GetLineSegment(line);
+    /// <summary>
+    ///  Matches related removed and added lines in a consecutive block of a patch document and marks identical parts dimmed.
+    /// </summary>
+    private void MarkInlineDifferences(IDocument document)
+    {
+        int index = 0;
+        List<DiffLineInfo> diffLines = [.. _diffLinesInfo.DiffLines.Values.OrderBy(i => i.LineNumInDiff)];
+        bool found = false;
 
-        document.MarkerStrategy.AddMarker(new TextMarker(lineSegment.Offset,
-            (endLine.Offset + endLine.TotalLength) -
-            lineSegment.Offset, TextMarkerType.SolidBlock, color,
-            ColorHelper.GetForeColorForBackColor(color)));
+        const int diffContentOffset = 1; // in order to skip the prefixes '-' / '+'
+        MarkerStrategy markerStrategy = document.MarkerStrategy;
 
-        return;
+        // Process the next blocks of removed / added diffLines and mark in-line differences
+        while (index < diffLines.Count)
+        {
+            found = false;
 
-        bool DoesLineStartWith(IDocument document, int offset, string prefixStr, bool invertMatch)
-            => invertMatch ^ LinePrefixHelper.DoesLineStartWith(document, offset, prefixStr);
-    }
+            // git-diff presents the removed lines directly followed by the added in a "block"
+            IReadOnlyList<ISegment> linesRemoved = GetBlockOfLines(diffLines, DiffLineType.Minus, ref index, ref found);
+            IReadOnlyList<ISegment> linesAdded = GetBlockOfLines(diffLines, DiffLineType.Plus, ref index, ref found);
 
-    private void SetText(ref string text)
-    {
-        if (!_useGitColoring)
-        {
-            return;
+            foreach (TextMarker marker in GetDifferenceMarkers(GetText, linesRemoved, linesAdded, diffContentOffset))
+            {
+                markerStrategy.AddMarker(marker);
+            }
         }
 
-        StringBuilder sb = new(text.Length);
-        AnsiEscapeUtilities.ParseEscape(text, sb, _textMarkers);
+        return;
 
-        text = sb.ToString();
+        string GetText(ISegment line)
+        {
+            int len = line.Length - diffContentOffset;
+            if (line.Offset + line.Length >= document.TextLength)
+            {
+                // This is likely a test, where last line do not have a newline
+                --len;
+            }
```

Comment:
Where does this come from?
I guess before this PR, `MarkInlineDifferences` did not see line feeds.
Perhaps, invert the condition?
Otherwise, `Math.Min` would be safer.

Reply:
I have guessed / remembered correctly. This PR _does_ change the behavior: Line feeds are fed to `AddDifferenceMarkers(List<TextMarker> markers, Func<ISegment, string> getText, ISegment lineRemoved, ISegment lineAdded, int beginOffset)` now.
The follow-up PR does not really change the behavior of this function. It mainly adapts to `string text` as replacement for `document`.

---

I do not understand neither what this condition shall achieve nor the comment. Where does it come from?

Reply:
There is a newline in patches that Git creates, the pesky \no newline.
Some test cases do not have the newline

Reply:
Why should we skip the last char if the newline is missing?!?

`line.Offset + line.Length == document.TextLength` is a valid situation. There is no need for adapting `len`.
`line.Offset + line.Length > document.TextLength` means `line` is kind of invalid regarding `document`. Use `Math.Min` in order to be safe because `--len` might not be enough.

In the follow-up it becomes:
```
        string GetText(ISegment line)
        {
            int len = line.Length - beginOffset;
            if (line.Offset + len == text.Length)
            {
                // This is likely a test, where last line do not have a newline
                --len;
            }

            return text.Substring(line.Offset + beginOffset, len);
        }
```
Same question.

Reply:
The core issue was in DiffLineNumAnalyzer.cs, could remove this fix

Reply:
> The core issue was in DiffLineNumAnalyzer.cs, could remove this fix

better
I took the liberty to push a proper support of CRLF.